btcec/field.go

```diff
@@ -226,20 +226,37 @@ func (f *fieldVal) SetBytes(b *[32]byte) *fieldVal {
 	return f
 }
 
-// SetByteSlice packs the passed big-endian value into the internal field value
-// representation.  Only the first 32-bytes are used.  As a result, it is up to
-// the caller to ensure numbers of the appropriate size are used or the value
-// will be truncated.
+// SetByteSlice interprets the provided slice as a 256-bit big-endian unsigned
+// integer (meaning it is truncated to the first 32 bytes), packs it into the
+// internal field value representation, and returns whether or not the resulting
+// truncated 256-bit integer is greater than or equal to the field prime (aka it
+// overflowed) in constant time.
+//
+// Note that since passing a slice with more than 32 bytes is truncated, it is
+// possible that the truncated value is less than the field prime.  It is up to
+// the caller to decide whether it needs to provide numbers of the appropriate
+// size or if it is acceptable to use this function with the described
+// truncation and overflow behavior.
 //
 // The field value is returned to support chaining.  This enables syntax like:
 // f := new(fieldVal).SetByteSlice(byteSlice)
+//
+// Preconditions: None
+// Output Normalized: Yes if no overflow, no otherwise
+// Output Max Magnitude: 1
 func (f *fieldVal) SetByteSlice(b []byte) *fieldVal {
 	var b32 [32]byte
-	for i := 0; i < len(b); i++ {
-		if i < 32 {
-			b32[i+(32-len(b))] = b[i]
-		}
+
+	if size := len(b); size >= 32 {
```

Comment:
Since you aren't constant time here anyway due to the branch, this can be simplified to:

```Go
	var b32 [32]byte
	if len(b) > 32 {
		b = b[:32]
	}
	copy(b32[32-len(b):], b)
	return f.SetBytes(&b32)
}
```

Reply:
There was no noticeable change in the benchmark (see the updated PR description). Here's the benchmarking code I used:

```go
func BenchmarkFieldVal_SetByteSlice(b *testing.B) {
	for n := 0; n < b.N; n++ {
		inBytes := hexToBytes("01fffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffff85c")
		var f fieldVal
		f.SetByteSlice(inBytes)
	}
}
```

I do think your version is better due to its simplicity.

Reply:
That's likely because your benchmark is being completely dominated by the hex conversion.  If you do a benchmark that only measures setting the slice, which is what you want to measure, you should see the difference.

e.g.
```Go
func BenchmarkFieldValSetByteSlice(b *testing.B) {
	inBytes := hexToBytes("01fffffffffffffffffffffffffffffffffffffffffffffffffffffffdfffff85c")
	for n := 0; n < b.N; n++ {
		var f fieldVal
		f.SetByteSlice(inBytes)
	}
}
```

Reply:
It was indeed the issue! The new benchmark result is down to `14.60 ns/op`, from `40.60 ns/op`! Thanks. 🙂